crates/solver/src/liquidity/uniswap_v3.rs

```diff
@@ -0,0 +1,286 @@
+use super::{slippage, AmmOrderExecution, ConcentratedLiquidity, LimitOrder, SettlementHandling};
+use crate::{
+    interactions::{
+        allowances::{AllowanceManager, AllowanceManaging, Allowances, Approval},
+        ExactOutputSingleParams, UniswapV3Interaction,
+    },
+    settlement::SettlementEncoder,
+};
+use anyhow::{ensure, Context, Result};
+use contracts::{GPv2Settlement, UniswapV3SwapRouter};
+use model::TokenPair;
+use num::rational::Ratio;
+use primitive_types::{H160, U256};
+use shared::{baseline_solver::BaseTokens, sources::uniswap_v3::pool_fetching::PoolFetching, Web3};
+use std::{collections::HashSet, ops::Mul};
+use std::{
+    sync::{Arc, Mutex},
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+// 1h timeout for Uniswap V3 interactions
+const TIMEOUT: u64 = 3600;
+
+pub struct UniswapV3Liquidity {
+    inner: Arc<Inner>,
+    pool_fetcher: Arc<dyn PoolFetching>,
+    settlement_allowances: Box<dyn AllowanceManaging>,
+    base_tokens: Arc<BaseTokens>,
+}
+pub struct Inner {
+    router: UniswapV3SwapRouter,
+    gpv2_settlement: GPv2Settlement,
+    // Mapping of how much allowance the router has per token to spend on behalf of the settlement contract
+    allowances: Mutex<Allowances>,
+}
+
+pub struct UniswapV3SettlementHandler {
+    inner: Arc<Inner>,
+    fee: Option<u32>,
+}
+
+/// Highly corelated to Uniswap V3 only.
+/// Converts:
+/// 1% fee to 10000
+/// 0.3% fee to 3000
+/// 0.05% to 500
+/// 0.01% to 100
+fn ratio_to_u32(ratio: Ratio<u32>) -> u32 {
+    ratio.mul(1_000_000).to_integer()
+}
```

Comment:
I'm worried this might panic. Could you check the mul implementation and see if there is a checked_mul?

Reply:
Switched to `checked_mul`. Thanks.